pkg/services/dashboards/dashboard_service.go

```diff
@@ -241,6 +245,33 @@ func (dr *dashboardServiceImpl) SaveDashboard(dto *SaveDashboardDTO) (*models.Da
 	return cmd.Result, nil
 }
 
+// DeleteDashboard removes dashboard from the DB. Errors out if the dashboard was provisioned. Should be used for
+// operations by the user where we want to make sure user does not delete provisioned dashboard.
+func (dr *dashboardServiceImpl) DeleteDashboard(dashboardId int64, orgId int64) error {
+	return dr.deleteDashboard(dashboardId, orgId, true)
+}
+
+// DeleteProvisionedDashboard removes dashboard from the DB even if it is provisioned.
+func (dr *dashboardServiceImpl) DeleteProvisionedDashboard(dashboardId int64, orgId int64) error {
+	return dr.deleteDashboard(dashboardId, orgId, false)
+}
+
+func (dr *dashboardServiceImpl) deleteDashboard(dashboardId int64, orgId int64, validateProvisionedDashboard bool) error {
+	if validateProvisionedDashboard {
+		isDashboardProvisioned := &models.IsDashboardProvisionedQuery{DashboardId: dashboardId}
+		err := bus.Dispatch(isDashboardProvisioned)
+		if err != nil {
+			return fmt.Errorf("error while checking if dashboard is provisioned: %v", err)
```

Comment:
Looks like I forgot to mention this line in my last review.

I think we can find this type of error creation in many places in Grafanas codebase but it's a bad idea since it makes it very hard to check for a type of error if it contains dynamic content. It's in general discouraged to include paths etc in error due to this reason. That said I also find this frustration in Go since passing that kind of information usually is very valuable. The solutions for this is either to log the dynamic information before returning it or creating a new error struct that can contain more information that the caller can typecast to and use.

Reply:
Not sure I follow here. My main point and frustration was that there is no stack trace and so was hard to find out where the error actually originated. I seen this style mentioned in some "best practices" kind of articles and it basically just chains the errors so you get something like `provision err: dashboard provisioning error: failed to open file: file does not exists`.

In regards to actually handling the error opposed to just logging them, I would assume you need to use sentinel kind of error with specific types that you can check but those I assume cannot be chained and so if they are not handled and just logged at the top, you loose the trace. Also checking just on the message itself, even if the error is not chained like here does not seem like a good idea.

So I wonder if not using `fmt.Errorf("current error string: %v", prevError)` would help here. It seems to me like creating specific error types for everything is not feasible (seems like a lot of code, especially when you do not want to handle most of the errors)?

While googling I seen this package https://github.com/pkg/errors which seems to work in a way to satisfy both constraints 1. allows for maintaining some kind of trace 2. keeping the type info so you can use sentinel error in the trace and handle them as needed. Would using something like that make sense?

Reply:
Sorry for misunderstanding your root problem. Yes, using `/pkg/errors` is totally fine with me. We should use that more. It's already vendored in Grafana ( as a sub dep, we should make it a first class dependency ). `/pkg/errors` will be a part of the stdlib eventually.

Reply:
I added the pkg/errors. It is pretty nifty. It can even print stacktrace with `%+v` which I did not know before.